File: backend/src/relocation_scout/api/actions.py
from __future__ import annotations

import json

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from relocation_scout.api.models import ActionResponse, ApprovalRequest
from relocation_scout.dependencies import get_session
from relocation_scout.persistence.unit_of_work import UnitOfWork
from relocation_scout.tools.approval_gateway import approval_gateway
from relocation_scout.workflow.controller import WorkflowController

router = APIRouter(tags=["actions"])


@router.get("/api/searches/{search_id}/actions", response_model=list[ActionResponse])
async def list_actions(search_id: str, session: AsyncSession = Depends(get_session)):
    uow = UnitOfWork(session)
    records = await uow.pending_actions.get_by_search(search_id)

    return [
        ActionResponse(
            action_id=r.id,
            action_type=r.action_type,
            target_listing_id=r.target_listing_id,
            payload=json.loads(r.payload_json),
            payload_hash=r.payload_hash,
            idempotency_key=r.idempotency_key,
            risk_level=r.risk_level,
            status=r.status,
            created_at=r.created_at.isoformat(),
        )
        for r in records
    ]


@router.get("/api/actions/{action_id}", response_model=ActionResponse)
async def get_action(action_id: str, session: AsyncSession = Depends(get_session)):
    uow = UnitOfWork(session)
    record = await uow.pending_actions.get(action_id)
    if not record:
        raise HTTPException(404, "Action not found")

    return ActionResponse(
        action_id=record.id,
        action_type=record.action_type,
        target_listing_id=record.target_listing_id,
        payload=json.loads(record.payload_json),
        payload_hash=record.payload_hash,
        idempotency_key=record.idempotency_key,
        risk_level=record.risk_level,
        status=record.status,
        created_at=record.created_at.isoformat(),
    )


@router.patch("/api/actions/{action_id}")
async def edit_action(
    action_id: str,
    body: dict,
    session: AsyncSession = Depends(get_session),
):
    uow = UnitOfWork(session)
    record = await uow.pending_actions.get(action_id)
    if not record:
        raise HTTPException(404, "Action not found")

    new_payload = body.get("payload", json.loads(record.payload_json))
    await approval_gateway.edit_action(record, new_payload)
    await session.commit()

    return ActionResponse(
        action_id=record.id,
        action_type=record.action_type,
        target_listing_id=record.target_listing_id,
        payload=new_payload,
        payload_hash=record.payload_hash,
        idempotency_key=record.idempotency_key,
        risk_level=record.risk_level,
        status="draft",
        created_at=record.created_at.isoformat(),
    )


@router.post("/api/actions/{action_id}/approve")
async def approve_action(
    action_id: str,
    body: ApprovalRequest,
    session: AsyncSession = Depends(get_session),
):
    uow = UnitOfWork(session)
    record = await uow.pending_actions.get(action_id)
    if not record:
        raise HTTPException(404, "Action not found")

    if record.status == "approved":
        raise HTTPException(400, "Action already approved")

    decision = await approval_gateway.approve(session, record, body.approved_by, body.comment)
    await session.commit()

    return {
        "approval_id": decision.approval_id,
        "action_id": action_id,
        "decision": "approved",
        "approved_by": body.approved_by,
        "approved_at": decision.approved_at.isoformat(),
    }


@router.post("/api/actions/{action_id}/reject")
async def reject_action(
    action_id: str,
    body: ApprovalRequest,
    session: AsyncSession = Depends(get_session),
):
    uow = UnitOfWork(session)
    record = await uow.pending_actions.get(action_id)
    if not record:
        raise HTTPException(404, "Action not found")

    decision = await approval_gateway.reject(session, record, body.approved_by, body.comment)
    await session.commit()

    return {
        "approval_id": decision.approval_id,
        "action_id": action_id,
        "decision": "rejected",
        "approved_by": body.approved_by,
    }


@router.post("/api/actions/{action_id}/execute")
async def execute_action(
    action_id: str,
    session: AsyncSession = Depends(get_session),
):
    uow = UnitOfWork(session)
    record = await uow.pending_actions.get(action_id)
    if not record:
        raise HTTPException(404, "Action not found")

    if record.status != "approved":
        raise HTTPException(
            400, f"Action must be approved before execution (current: {record.status})"
        )

    controller = WorkflowController(session)
    try:
        result = await controller.execute_approved_action(record.search_id, action_id)
        return {"status": "executed", "result": result}
    except Exception as e:
        raise HTTPException(500, f"Action execution failed: {e!s}") from e

File: frontend/src/api/actions.ts
import { api } from './client';
import type { Action } from '../types';

type BackendActionResponse = {
  action_id: string;
  action_type: string;
  target_listing_id: string;
  payload: Record<string, unknown>;
  payload_hash: string;
  idempotency_key: string;
  risk_level: string;
  status: string;
  created_at: string | null;
};

type ActionUpdateRequest = {
  payload: {
    recipient: string;
    subject: string;
    body: string;
  };
};

function mapActionType(actionType: string): Action['action_type'] {
  if (
    actionType === 'email_realtor' ||
    actionType === 'phone_call' ||
    actionType === 'book_viewing' ||
    actionType === 'send_inquiry'
  ) {
    return actionType;
  }

  return 'send_inquiry';
}

function mapActionStatus(status: string): Action['status'] {
  if (status === 'draft') {
    return 'pending';
  }
  if (
    status === 'pending' ||
    status === 'approved' ||
    status === 'rejected' ||
    status === 'executing' ||
    status === 'completed' ||
    status === 'failed'
  ) {
    return status;
  }

  return 'pending';
}

function mapRiskLevel(level: string): Action['risk_level'] {
  if (level === 'low' || level === 'medium' || level === 'high') {
    return level;
  }
  return 'low';
}

function parseString(value: unknown): string {
  return typeof value === 'string' ? value : '';
}

function toFrontendAction(
  backendAction: BackendActionResponse,
  searchId: string,
): Action {
  const payload = backendAction.payload ?? {};
  const recipient = parseString(payload.recipient);
  const subject = parseString(payload.subject);
  const body = parseString(payload.body);
  const listingTitleFallback = backendAction.target_listing_id
    ? `Listing ${backendAction.target_listing_id.slice(0, 8)}`
    : 'Listing';
  const createdAt = backendAction.created_at ?? new Date().toISOString();

  return {
    id: backendAction.action_id,
    search_id: searchId,
    action_type: mapActionType(backendAction.action_type),
    listing_id: backendAction.target_listing_id,
    listing_title: subject || listingTitleFallback,
    recipient,
    subject,
    body,
    risk_level: mapRiskLevel(backendAction.risk_level),
    status: mapActionStatus(backendAction.status),
    payload_hash: backendAction.payload_hash,
    idempotency_key: backendAction.idempotency_key,
    created_at: createdAt,
    updated_at: createdAt,
    approved_at: null,
    executed_at: null,
    failure_reason: null,
  };
}

export const actionsApi = {
  list: (searchId: string) =>
    api
      .get<BackendActionResponse[]>(`/api/searches/${searchId}/actions`)
      .then((actions) => actions.map((action) => toFrontendAction(action, searchId))),

  get: (id: string, searchId = '') =>
    api
      .get<BackendActionResponse>(`/api/actions/${id}`)
      .then((action) => toFrontendAction(action, searchId)),

  update: (id: string, data: ActionUpdateRequest) =>
    api.patch<BackendActionResponse>(`/api/actions/${id}`, data),

  approve: (id: string, approvedBy = 'demo_user', comment?: string) =>
    api.post(`/api/actions/${id}/approve`, { approved_by: approvedBy, comment }),

  reject: (id: string, approvedBy = 'demo_user', comment?: string) =>
    api.post(`/api/actions/${id}/reject`, { approved_by: approvedBy, comment }),

  execute: (id: string) =>
    api.post(`/api/actions/${id}/execute`),
};
